google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -358,39 +352,35 @@ class ResumableAsyncStreamingReadWriteRpcImpl
   }
 
   void ReadWriteRetryFailedStream() {
-    promise<void> root;
-    {
-      std::lock_guard<std::mutex> g{mu_};
-      if (stream_state_ != State::kInitialized) return;
-
-      stream_state_ = State::kRetrying;
-      assert(!retry_promise_.has_value());
-      retry_promise_.emplace();
-
-      // Assuming that a `Read` fails:
-      // If an outstanding operation is present, we can't enter the retry
-      // loop, so we defer it until the outstanding `Write` finishes at
-      // which point we can enter the retry loop. Since we will return
-      // `reinit_done_`, we guarantee that another operation of the same type
-      // is not called while we're waiting for the outstanding operation to
-      // finish and the retry loop to finish afterward.
-
-      future<void> root_future = root.get_future();
-
-      // at most one of these will be set
-      if (in_progress_read_.has_value()) {
-        root_future =
-            root_future.then(ChainFuture(in_progress_read_->get_future()));
-      }
-      if (in_progress_write_.has_value()) {
-        root_future =
-            root_future.then(ChainFuture(in_progress_write_->get_future()));
-      }
-
-      root_future.then([this](future<void>) { FinishOnStreamFail(); });
+    AsyncRoot root;
```

Comment:
Please use a separate PR for these `AsyncRoot` refactorings.

Reply:
https://github.com/googleapis/google-cloud-cpp/pull/8505